Add state system

gmanjames opened the pull request:
I added a general state system that can be used to create states for generic FlxObject. I extended the BaseState class that is at the core of this system to create a PlayerState class that can be extended to create player related states. Further, I extended this PlayerState class to create some basic player states that handle walking, running and jumping.

gmanjames committed:
Added state system.

Added a BaseState class to be extended to create states for any FlxObject, and a PlayerState class to be extended to create player related states. Added basic player related states that handle walking running and jumping.

gmanjames committed:
Player deccelerates when movement key is released while on ground

gmanjames committed:
Cleanup and tuning.

Removed not used method from player class. Keep x.velocity when directional keys are released while jumping.

gmanjames committed:
Added documentation

gmanjames committed:
Added FSM class

gmanjames committed:
Corrected whitespace issue

gmanjames commented:
Also I corrected the issues with the funky whitespace

gmanjames committed:
Cleanup. Whitespace issue within method

SamBumgardner commented:
Cool, I'll give this a proper review before class starts!

SamBumgardner commented on source/states/BaseState.hx:
```diff
@@ -0,0 +1,96 @@
+{
+  public  var nextState:Null<BaseState>;
+
+  private var duration:Int;
```
Since we're using acceleration and velocity in this game, I think we might want to move away from measuring the duration of a state in frames. Or maybe not. I'm pretty torn, because I think animations in HaxeFlixel are still done on a frame-by-frame basis.

Do you think it would be too confusing to have two duration variables, one that's an int measuring frames, and the other that's a float measuring time (in seconds)?

You could use a similar system where -1 signals that the duration isn't measured with that variable, and otherwise if either runs out, the nextState is activated.

gmanjames replied:
I don't think that would be difficult to implement. With proper naming I don't think that would be too confusing either. There may be certain states that need to have a measure of "framesLeft" as well as a time based duration value as well. I will make it happen

SamBumgardner commented on source/states/PlayerAirState.hx:
```diff
@@ -0,0 +1,94 @@
+      }
+    }
+
+    // Slow down if no direction held
```
This comment is a bit misleading (and I'm pretty sure I'm the one that wrote it, sorry). The player doesn't slow down in the air if no direction is held, they just stop accelerating and maintain their current momentum.

SamBumgardner commented on source/states/PlayerState.hx:
```diff
@@ -0,0 +1,71 @@
+   *
+   * @return scalar value of the players next horizontal move
+   */
+  private function pollForHorizontalMove():Int
```
Two things: 
 1. you might want to have a function to poll for other sorts of inputs, like jumping
 2. I feel like it may be more appropriate to include this stuff in the Player class, since I feel that the finite state machine is focused on interpreting inputs rather than getting them in the first place.

SamBumgardner commented on source/states/PlayerState.hx:
```diff
@@ -0,0 +1,71 @@
+   public function isPlayerRunning():Bool
+   {
+     return FlxG.keys.anyPressed([FlxKey.Z]);
+   }
```
See my above note on the horizontal movement polling.

SamBumgardner reviewed (changes requested):
The code looks good to me and I didn't run into any problems during testing. I just have a few comments on minor changes/adjustments that I think should be made.

Let me know what you think, and we'll figure out what actually needs to change and what's probably fine to leave as-is.

gmanjames committed:
Pull request #5 comments.

Give user option to use time based duration or framesLeft in state. Moved some player related methods to Player class.

gmanjames commented:
I have corrected some of the things discussed in class. Can now allow the user to choose between a time based duration or a frames left option. Also I moved some player related methods over to the Player class and deleted one state that became unnecessary as a result.

gmanjames committed:
Added variable height jump

gmanjames commented:
I will probably switch to a new branch soon as these latest changes are starting to get away from the basic addition of a state system. Added a variable height jump that can certainly be played around with to find a good balance.

SamBumgardner commented:
I'll review your most recent changes before class and (unless there's some weird problem that comes up) merge it in!

gmanjames committed:
Added documentation for undocumented method

gmanjames committed:
Tweaks to player jump.

Jump degrades based off of velocity instead of time. Removed some magic numbers

SamBumgardner commented:
The code looks good!

I think some systems could still use tweaking (prevent the player from jumping automatically if jump button is held, introduce a buffer window for jump inputs, smooth out jump start up) but the main goals of this branch have been accomplished. That, and other people depend on the player class so merging in this branch should make everyone else's lives easier.

One optional change that I'd like is adding some documentation to the different state child classes. I don't think it's strictly necessary because HaxeDevelop will automatically pop up the parent class' documentation when autocompleting the child's functions, but more specific documentation could help clarify what a specific state's functions are meant to accomplish.

If you want to  document that, go for it, otherwise we're good to merge!

SamBumgardner commented on source/states/PlayerFallState.hx:
```diff
@@ -0,0 +1,40 @@
+import flixel.FlxObject;
+import flixel.input.keyboard.FlxKey;
+
+class PlayerFallState extends PlayerGroundState
```
`PlayerFallState` should extend `PlayerAirState`, not `PlayerGroundState`.

SamBumgardner commented on source/states/PlayerAirState.hx:
```diff
@@ -0,0 +1,103 @@
+  {
+      var player = cast(object, Player);
+      player.velocity.y = -JUMP;
+  }
```
I don't think that the PlayerAirState's default behavior should be to make the character jump. Any action that makes the player lose contact with the ground will put them in the air state, which goes beyond just jumping.

SamBumgardner commented on source/states/PlayerJumpState.hx:
```diff
@@ -0,0 +1,28 @@
+import flixel.FlxObject;
+import flixel.input.keyboard.FlxKey;
+
+class PlayerJumpState extends PlayerAirState
```
Put jump enter logic in this subclass instead of in PlayerAirState.

SamBumgardner commented on source/states/PlayerGroundState.hx:
```diff
@@ -0,0 +1,108 @@
+      nextState = new PlayerJumpState();
+      return true;
+    }
+
```
Also need to check if the player isn't in contact with the ground, put them in the appropriate air state. Probably do an else if () to put them in fall state if they aren't jumping.

SamBumgardner reviewed (changes requested):
I found a few issues when looking over your code. Fixing these should clear up the issues I noticed in gameplay.

gmanjames committed:
Fixed some noted issues

Moved some logic to jumpstate. Changed an inheritance chain
